### PR TITLE
Engine Refactor

### PR DESCRIPTION
**Pull request by benbjohnson**
## Overview

This pull request extracts the storage concerns from the `tsdb` package and moves them in to the `tsdb/engine/v1` package. This allows us to support multiple storage formats inside of InfluxDB and it allows us to incrementally migrate between them.

One consequence of this refactor is that the `tsdb` tests needed to be moved to their own `tsdb_test` package so that both the `tsdb` and `engine/v1` packages could be imported without causing a circular dependency.

This engine refactor is being PR'd separately from the compression engine.

### CODE REVIEW

File: tsdb/engine.go [hunk 1]
```diff
@@ -1,892 +1,127 @@
...
 
-	if (e.stmt.IsRawQuery && !e.stmt.HasDistinct()) || e.stmt.IsSimpleDerivative() {
-		go e.executeRaw(out)
-	} else {
-		go e.executeAggregate(out)
-	}
-	return out
-}
+	Begin(writable bool) (Tx, error)
+	DB() *bolt.DB
```
Comment (dgnorton) [inline tsdb/engine.go]: Should this return an `interface{}`?

File: tsdb/engine.go [hunk 2]
```diff
@@ -1,892 +1,127 @@
...
 
-// mappersDrained returns whether all the executors Mappers have been drained of data.
-func (e *Executor) mappersDrained() bool {
-	for _, m := range e.mappers {
-		if !m.drained {
-			return false
-		}
-	}
-	return true
+	Flush(partitionFlushDelay time.Duration) error
```
Comment (dgnorton) [inline tsdb/engine.go]: Flushing the WAL seems like an implementation detail that should be hidden.  I know we have code outside the engine that uses it now (e.g., on startup) but maybe we shouldn't.  Does the WAL really need to be flushed on startup or could the engine just handle it under the covers in its normal flush cycle?

File: tsdb/engine.go (hunk 1 above)
Comment (benbjohnson) [inline tsdb/engine.go]: I actually only needed this temporarily while refactoring. Removed in 21fa72d.

File: tsdb/engine.go (hunk 2 above)
Comment (benbjohnson) [inline tsdb/engine.go]: This was only being used by a test so I was able to remove it from the interface: 0c61e20.

**Comment by benbjohnson** [comment]
OK, I rebased from @otoolep's changes and everything seems to be in working order.

File: tsdb/executor.go [hunk 3]
```diff
@@ -0,0 +1,892 @@
...
+				selectNames: e.stmt.NamesInSelect(),
+				fields:      e.stmt.Fields,
+				c:           out,
+			}
+		}
+		if e.stmt.HasDerivative() {
+			interval, err := derivativeInterval(e.stmt)
+			if err != nil {
+				out <- &influxql.Row{Err: err}
+				return
```
Comment (jwilder) [inline tsdb/executor.go]: Might not be part of this PR but this early return looks like it will not close the out channel on L292.

Comment (otoolep) [inline tsdb/executor.go]: Good catch. I wonder if deferring `close(out)` would be a sensible thing to do here?

**Comment by jwilder** [comment]
Nice refactoring.  LGTM :+1:
